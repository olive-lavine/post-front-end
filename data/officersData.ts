import { Officer } from "../types";
// import { getConnection } from 'typeorm';

const officersData: Officer[] = [
  {
    name: "John Doe",
    id: "123",
    state: "WA",
    agencyName: "Seattle Police Department",
    startDate: "2015-01-15",
    endDate: "2020-05-30",
    separationReason: "Retirement",
  },
  {
    name: "Jane Smith",
    id: "456",
    state: "WA",
    agencyName: "Bellevue Police Department",
    startDate: "2018-03-22",
    endDate: "2022-07-18",
    separationReason: "Resignation",
  },
  {
    name: "Alice Johnson",
    id: "789",
    state: "WA",
    agencyName: "Tacoma Police Department",
    startDate: "2019-11-01",
    endDate: "2023-02-28",
    separationReason: "Termination",
  },
  {
    name: "Bob Johnson",
    id: "987",
    state: "WA",
    agencyName: "Spokane Police Department",
    startDate: "2020-08-10",
    endDate: "2024-12-31",
    separationReason: "Transfer",
  },
  {
    name: "Sarah Davis",
    id: "654",
    state: "WA",
    agencyName: "Vancouver Police Department",
    startDate: "2017-06-05",
    endDate: "2021-09-15",
    separationReason: "Promotion",
  },
  {
    name: "Josh Doe",
    id: "223",
    state: "WA",
    agencyName: "Seattle Police Department",
    startDate: "2015-01-15",
    endDate: "2020-05-30",
    separationReason: "Retirement",
  },
  {
    name: "Jane Doe",
    id: "124",
    state: "WA",
    agencyName: "Seattle Police Department",
    startDate: "2015-01-15",
    endDate: "2020-05-30",
    separationReason: "Retirement",
  },
  {
    name: "Emily Johnson",
    id: "345",
    state: "WA",
    agencyName: "Bellevue Police Department",
    startDate: "2020-02-12",
    endDate: "2024-06-30",
    separationReason: "Resignation",
  },
  {
    name: "Michael Smith",
    id: "567",
    state: "WA",
    agencyName: "Tacoma Police Department",
    startDate: "2016-09-18",
    endDate: "2020-12-31",
    separationReason: "Termination",
  },
  {
    name: "Olivia Davis",
    id: "890",
    state: "WA",
    agencyName: "Spokane Police Department",
    startDate: "2017-12-01",
    endDate: "2022-03-31",
    separationReason: "Transfer",
  },
  {
    name: "William Johnson",
    id: "432",
    state: "WA",
    agencyName: "Vancouver Police Department",
    startDate: "2019-03-10",
    endDate: "2023-08-15",
    separationReason: "Promotion",
  },
  {
    name: "Alex Brown",
    id: "111",
    state: "WA",
    agencyName: "Seattle Police Department",
    startDate: "2014-02-28",
    endDate: "2019-06-15",
    separationReason: "Retirement",
  },
  {
    name: "Emma Johnson",
    id: "222",
    state: "WA",
    agencyName: "Bellevue Police Department",
    startDate: "2017-09-10",
    endDate: "2021-12-31",
    separationReason: "Resignation",
  },
  {
    name: "Daniel Smith",
    id: "333",
    state: "WA",
    agencyName: "Tacoma Police Department",
    startDate: "2018-05-15",
    endDate: "2022-08-30",
    separationReason: "Termination",
  },
  {
    name: "Sophia Davis",
    id: "444",
    state: "WA",
    agencyName: "Spokane Police Department",
    startDate: "2019-10-01",
    endDate: "2023-01-31",
    separationReason: "Transfer",
  },
  {
    name: "Ethan Johnson",
    id: "555",
    state: "WA",
    agencyName: "Vancouver Police Department",
    startDate: "2016-03-10",
    endDate: "2020-08-15",
    separationReason: "Promotion",
  },
  {
    name: "Mia Smith",
    id: "666",
    state: "WA",
    agencyName: "Seattle Police Department",
    startDate: "2015-07-20",
    endDate: "2020-11-30",
    separationReason: "Retirement",
  },
  {
    name: "Noah Davis",
    id: "777",
    state: "WA",
    agencyName: "Bellevue Police Department",
    startDate: "2018-10-05",
    endDate: "2023-01-15",
    separationReason: "Resignation",
  },
  {
    name: "Ava Johnson",
    id: "888",
    state: "WA",
    agencyName: "Tacoma Police Department",
    startDate: "2019-12-20",
    endDate: "2024-04-30",
    separationReason: "Termination",
  },
  {
    name: "Liam Smith",
    id: "999",
    state: "WA",
    agencyName: "Spokane Police Department",
    startDate: "2020-06-01",
    endDate: "2024-09-30",
    separationReason: "Transfer",
  },
  {
    name: "Isabella Davis",
    id: "000",
    state: "WA",
    agencyName: "Vancouver Police Department",
    startDate: "2017-01-10",
    endDate: "2021-04-15",
    separationReason: "Promotion",
  },
];
// export const getOfficersByStateAndAgency = async (state: string, agency: string): Promise<Officer[]> => {
//   const connection = await getConnection(); // Assuming you have already set up a database connection using TypeORM

//   const officerRepository = connection.getRepository(Officer); // Assuming you have defined an Officer entity in TypeORM

//   let query = officerRepository.createQueryBuilder('officer');

//   if (state !== 'all') {
//     query = query.where('officer.state = :state', { state });
//   }

//   if (agency !== 'all') {
//     query = query.andWhere('officer.agencyName = :agency', { agency });
//   }

//   return query.getMany();
// };
export const getOfficersByStateAndAgency = (
  state: string,
  agency: string,
  page: number,
  perPage: number
): { officers: Officer[]; totalItems: number } => {
  // Filter the data based on state and agency
  const filteredOfficers = officersData.filter(
    (officer) =>
      (state === "all" || officer.state === state) &&
      (agency === "all" || officer.agencyName === agency)
  );

  //calculate total items after filtering
  const totalItems = filteredOfficers.length;

  //calculate the range of items after filtering
  const startIndex = (page - 1) * perPage;
  const endIndex = Math.min(startIndex + perPage, totalItems);
  const paginatedOfficers = filteredOfficers.slice(startIndex, endIndex);

  return {
    officers: paginatedOfficers,
    totalItems,
  };
};
