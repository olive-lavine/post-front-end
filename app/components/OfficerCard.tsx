import React from "react";
import { PeaceOfficer, WorkHistory } from "../../data/officersData"; // Ensure correct import path

interface OfficerCardProps {
  officer: PeaceOfficer;
}

const OfficerCard: React.FC<OfficerCardProps> = ({ officer }) => {
  return (
    <div className="officer-card">
      <h3 className="standout officer-name">
        {officer.firstName} {officer.lastName}
      </h3>
      <p className="uid">UID: {officer.UID}</p>
      <div className="work-history">
        {officer.workHistory.length === 0 ? (
          <p>No work history available</p>
        ) : (
          officer.workHistory.map((history: WorkHistory) => (
            <div key={history.id} className="work-history-entry">
              <p className="agency-name">
                {history.agencyName || "Unknown Agency"}
              </p>
              <p>
                Start Date:{" "}
                {history.startDate
                  ? new Date(history.startDate).toLocaleDateString()
                  : ""}
              </p>
              <p>
                End Date:{" "}
                {history.separationDate
                  ? new Date(history.separationDate).toLocaleDateString()
                  : ""}
              </p>
              <p>Separation Reason: {history.separationReason}</p>
              <br></br>
              <br></br>
            </div>
          ))
        )}
      </div>
    </div>
  );
};

export default OfficerCard;
