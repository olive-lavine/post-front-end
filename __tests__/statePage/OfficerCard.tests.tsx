// OfficerCard.test.tsx
import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import OfficerCard from "../../app/components/OfficerCard";
import { PeaceOfficer, WorkHistory } from "../../data/officersData";

describe("OfficerCard", () => {
  const mockOfficer: PeaceOfficer = {
    id: 1,
    UID: "123456",
    firstName: "John",
    lastName: "Doe",
    workHistory: [
      {
        id: 1,
        startDate: "2021-12-30",
        separationDate: "2022-01-30",
        separationReason: "Resigned",
        agencyName: "Example Agency",
      },
    ],
  };

  it("should render officer details correctly", () => {
    render(<OfficerCard officer={mockOfficer} />);

    // Check if the officer's name is rendered
    expect(screen.getByText("John Doe")).toBeInTheDocument();

    // Check if the UID is rendered
    expect(screen.getByText("UID: 123456")).toBeInTheDocument();

    // Check if the work history is rendered correctly
    expect(screen.getByText("Example Agency")).toBeInTheDocument();

    // Flexible date check
    const startDateText = screen.getByText(/Start Date:/).textContent!;
    const endDateText = screen.getByText(/End Date:/).textContent!;

    // Ensure that the start and end dates are in some recognizable format
    expect(startDateText).toMatch(/Start Date: \d{1,2}\/\d{1,2}\/\d{4}/);
    expect(endDateText).toMatch(/End Date: \d{1,2}\/\d{1,2}\/\d{4}/);

    // Check for separation reason
    expect(screen.getByText("Separation Reason: Resigned")).toBeInTheDocument();
  });

  it('should display "No work history available" when work history is empty', () => {
    const emptyWorkHistoryOfficer: PeaceOfficer = {
      id: 2,
      UID: "789012",
      firstName: "Jane",
      lastName: "Smith",
      workHistory: [],
    };

    render(<OfficerCard officer={emptyWorkHistoryOfficer} />);

    // Check if the no work history message is displayed
    expect(screen.getByText("No work history available")).toBeInTheDocument();
  });

  it('should display "Unknown Agency" if agencyName is undefined or null', () => {
    const officerWithUndefinedAgency: PeaceOfficer = {
      id: 3,
      UID: "345678",
      firstName: "Emily",
      lastName: "Johnson",
      workHistory: [
        {
          id: 3,
          startDate: "2019-05-01",
          separationDate: "2021-05-01",
          separationReason: "Transferred",
          agencyName: undefined, // or null
        },
      ],
    };

    render(<OfficerCard officer={officerWithUndefinedAgency} />);

    // Check if "Unknown Agency" is displayed
    expect(screen.getByText("Unknown Agency")).toBeInTheDocument();
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(<OfficerCard officer={mockOfficer} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
