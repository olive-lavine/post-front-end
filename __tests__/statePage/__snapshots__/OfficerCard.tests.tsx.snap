// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfficerCard should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="officer-card"
  >
    <h3
      class="standout officer-name"
    >
      <br />
      John Doe
    </h3>
    <p
      class="uid"
    >
      UID: 123456
    </p>
    <div
      class="work-history"
    >
      <div
        class="work-history-entry"
      >
        <p
          class="agency-name"
        >
          Example Agency
        </p>
        <p>
          Start Date: 12/29/2021
        </p>
        <p>
          End Date: 1/29/2022
        </p>
        <p>
          Separation Reason: Resigned
        </p>
        <br />
        <br />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
